cilium, gops: remap to fixed port to avoid collision with nodeport range

borkmann opened the pull request:
See commit msg.

borkmann committed:
cilium, gops: remap to fixed port to avoid collision with nodeport range

Lee reported that kube-proxy log had a warning that its bind protection
couldn't bind a specific port in the nodeport range. Turns out gops was
using this particular port already through it's auto-binding (127.0.0.1:0).
Meaning that in case gops collides with a NodePort service, we might
not be able to pull gops data from that port since either kube-proxy or
kube-proxt free variant will redirect us to the actual service instead.

Given this is rather unpredictable wrt which port the agent will bind for
gops, remap it to a fixed default port and add a user configurable knob
that allows to use a different one if necessary. Given the agent, operator,
clustermesh-apiserver and hubble-relay all start the gops listener, add
the --gops-port flag to each of them. The CNI does not have gops enabled
by default but only in debug mode hence no changes there for now given
it's unlikely being used this way in production.

Fixes: #14218
Reported-by: Lee Hu via Slack
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

rolinh reviewed (changes requested):
Nice catch! Could you please also update Hubble Relay as I think it's also affected (nodeport protection is global)? It also enables gops by default with the default listen address (see `hubble-relay/cmd/serve/serve.go`).

aanm reviewed (changes requested):
the overall changes LGTM, however a couple points that we need to take care:
1) It would likely be better to `Fatal` if we can't initialize gops. If gops isn't running at all we won't be able to debug Cilium if we need it;
2) Currently, the default is `127.0.0.1:0` and we are changing it to `localhost:XXXX`, maybe we should keep the `127.0.0.1`.
3) I don't see a "reuse port" in gops, so it is likely that we will hit something similar as https://github.com/cilium/cilium/issues/11573.

rolinh commented on hubble-relay/cmd/serve/serve.go:
```diff
@@ -184,7 +189,8 @@ func runServe(vp *viper.Viper) error {
 	gopsEnabled := vp.GetBool(keyGops)
 	if gopsEnabled {
-		if err := agent.Listen(agent.Options{}); err != nil {
+		addr := fmt.Sprintf("localhost:%d", viper.GetInt(keyGopsPort))
```
This won't work as there is no global viper instance and thus the port will always be `0`. Please, use `vp` instead.
```suggestion
		addr := fmt.Sprintf("localhost:%d", vp.GetInt(keyGopsPort))
```

borkmann replied:
Indeed, thanks a lot for the catch!

rolinh reviewed (changes requested):
The Hubble Relay change is not correct, please see comment.

tklauser commented on operator/main.go:
```diff
@@ -70,10 +70,13 @@ var (
+			if err := gops.Listen(gops.Options{Addr: addr}); err != nil {
+				log.WithError(err).WithFields(addrField).Fatal("Cannot start gops server")
+			} else {
+				log.WithFields(addrField).Info("Started gops server")
```
Nit: the `else` branch here is not necessary given that we now `Fatal` on error.

```suggestion
			if err := gops.Listen(gops.Options{Addr: addr}); err != nil {
				log.WithError(err).WithFields(addrField).Fatal("Cannot start gops server")
			}
			log.WithFields(addrField).Info("Started gops server")
```

tklauser commented on daemon/cmd/daemon_main.go:
```diff
@@ -115,9 +115,12 @@ var (
+			if err := gops.Listen(gops.Options{Addr: addr}); err != nil {
+				log.WithError(err).WithFields(addrField).Fatal("Cannot start gops server")
+			} else {
+				log.WithFields(addrField).Info("Started gops server")
```
Nit: `else` branch here is not necessary given the `Fatal` above.

borkmann replied:
fixed

tklauser commented on clustermesh-apiserver/main.go:
```diff
@@ -68,6 +68,15 @@ var (
+			if err := gops.Listen(gops.Options{Addr: addr}); err != nil {
+				log.WithError(err).WithFields(addrField).Fatal("Cannot start gops server")
+			} else {
+				log.WithFields(addrField).Info("Started gops server")
```
Nit: `else` branch here is not necessary given the `Fatal` above.

borkmann replied:
fixed

tklauser reviewed (changes requested):
Implementation LGTM apart from the `else` branch nits.

Might be worth adding the new default ports to the table in `Documentation/operations/system_requirements.rst` as well (search for `The following ports should also be available on each node`, cannot link to the source file due to GitHub rendering it).

As already commented, will take care of adding `SO_REUSEADDR`/`SO_REUSEPORT` to `gops` upstream. We can then pull this in a follow-up PR.

tklauser reviewed (approved):
Thanks!

borkmann committed:
vendor: point gops to v0.3.14 tag for SO_REUSEPORT

For integrating work in: https://github.com/google/gops/pull/132

Resync'ed vendor directory to pull in latest gops via following:

  go get github.com/google/gops@latest && go mod vendor && go mod tidy && git add vendor/ go.mod go.sum && git commit -sa

Fixes #14332
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
docker: switch gops to v0.3.14

Switch all gops versions in the docker images from v0.3.10 to v0.3.14.
This is the client side though, but given we're outdated in our images
lets update as well to latest version.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

brb reviewed (approved):
Good find, LGTM.

tklauser reviewed (commented):
I think the first two commits should be merged so the version in `go.mod` is bumped as part of the same commit as the `vendor/` directory.

You may also want to add `Fixes #14332` to the commit message bumping `gops`.

Otherwise LGTM, thanks!

tklauser reviewed (approved):
LGTM with https://github.com/cilium/cilium/pull/14329#issuecomment-742597619, thanks!
